libpysal/graph/_kernel.py

```diff
@@ -183,33 +207,47 @@ def _kernel(
                 d = spatial.distance.pdist(coordinates, metric=metric, **dist_kwds)
                 sq = spatial.distance.squareform(d)
 
-            # ensure that self-distance is dropped but 0 between co-located pts not
             # get data and ids for sparse constructor
             data = sq.flatten()
             i = numpy.tile(numpy.arange(sq.shape[0]), sq.shape[0])
             j = numpy.repeat(numpy.arange(sq.shape[0]), sq.shape[0])
-            # remove diagonal
-            data = numpy.delete(data, numpy.arange(0, data.size, sq.shape[0] + 1))
-            i = numpy.delete(i, numpy.arange(0, i.size, sq.shape[0] + 1))
-            j = numpy.delete(j, numpy.arange(0, j.size, sq.shape[0] + 1))
+            # remove diagonal to ensure that self-distance is dropped
+            # but 0 between co-located pts not
+            # data = numpy.delete(data, numpy.arange(0, data.size, sq.shape[0] + 1))
+            # i = numpy.delete(i, numpy.arange(0, i.size, sq.shape[0] + 1))
+            # j = numpy.delete(j, numpy.arange(0, j.size, sq.shape[0] + 1))
             # construct sparse
             d = sparse.csc_array((data, (i, j)))
         else:
             d = sparse.csc_array(coordinates)
+
     if bandwidth is None:
         bandwidth = numpy.percentile(d.data, 25)
+
     elif bandwidth == "auto":
         if (kernel == "identity") or (kernel is None):
             bandwidth = numpy.nan  # ignored by identity
         else:
             bandwidth = _optimize_bandwidth(d, kernel)
+
     if callable(kernel):
-        d.data = kernel(d.data, bandwidth)
+        similarities = kernel(d.data, bandwidth)
+        k0 = kernel(0, bandwidth)
     else:
-        d.data = _kernel_functions[kernel](d.data, bandwidth)
-
-    if taper:
+        similarities = _kernel_functions[kernel](d.data, bandwidth)
+        k0 = _kernel_functions[kernel](0, bandwidth)
+
+    if taper is not False:
+        if taper is True:
+            taper = bandwidth
+        similarities[d.data > bandwidth] = 0
```

Comment:
I think this should be `similarities.min()`... just in case: the silverman kernel has negative values at large distances.

Reply:
Never mind! I'm torn---setting to zero explicitly means the links are fully removed by `d.eliminate_zeros()`, which is the original intent of tapering. Setting to `similarities.min()` means these links are retained, and thus not "tapered" strictly speaking.